### PR TITLE
34 fine tuning and skypilot

### LINKED ISSUE #34: Fine-tuning and skypilot
Try fine-tuning for the taxonomy classifier. This will not run locally, so try skypilot for accessing AWS resources.

### PR DESCRIPTION
**Pull request by RFOxbury**
---

# Description

This PR incorporates a lot of code from the previous Innovation Sweet Spots repo (see code [here](https://github.com/nestauk/innovation_sweet_spots/tree/dev/innovation_sweet_spots/analysis/notebooks/review_labelling)) and uses it to fine-tune DistilBERT for our multi-label use case.

It's still very scrappy but thought it was worth opening for review so that you two can see if you're also able to run a notebook using skypilot!

Fixes #34

# Instructions for Reviewer

In order to test the code in this PR you need to ...

- Follow the instructions in `infra/skypilot/README.md` to set up an AWS cluster using skypilot
- Open up `discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py' in jupyter notebook. Because jupytext is installed, you should be able to open it as a jupyter notebook.
- In that notebook, set the variable `PRODUCTION` to `False` and run all of the cells. All of these should run fine. This trains the transformer model on just 1000 datapoints.
- If you want to, you can also set `PRODUCTION` to `True` and run on more data!

Please pay special attention to ...

- Whether you can actually launch skypilot using the instructions in the README. If you can't, this means there's an issue with the instructions
- Whether the notebook runs and logs everything on weights and biases correctly
- Whether the explanation above is correct 🙃 Is it ok to say "... trains the transformer model ..." ? I guess what is happening is the transformer model is already initialised, and the `trainer.train()` step trains it on our data. But I am not sure I'm using the right language
- Whether the code and methods make sense generally. At this point, the transformer just runs for 3 epochs which I guess probably isn't enough? I tried to set the hyperparameters to values that would make the whole thing run quickly.

# Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

**Comment by j-gillam** [comment]
Hey @RFOxbury ! Skypilot launched perfectly with your instructions!! It did ask me for the token. 

I have had to do a couple of extra steps to get your script to run:
It wouldn't let me open the py file as an ipynb. I instead had to convert from py to ipynb, e.g. in the terminal type:
`jupytext --to notebook discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py`
and also needed to add the kernel so it would show up as an option in jupyterlab: 
`python -m ipykernel install --user --name=discovery_child_development`

Everything now seems to be running! I will now have a look at the code specifics :) 

Quick q about skypilot: 
Do we need to use jupyter notebooks? Could we instead just run the python files through the terminal we have SSH'd into?

**Comment by RFOxbury** [comment]
> Hey @RFOxbury ! Skypilot launched perfectly with your instructions!! It did ask me for the token.
> 
> I have had to do a couple of extra steps to get your script to run: It wouldn't let me open the py file as an ipynb. I instead had to convert from py to ipynb, e.g. in the terminal type: `jupytext --to notebook discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py` and also needed to add the kernel so it would show up as an option in jupyterlab: `python -m ipykernel install --user --name=discovery_child_development`
> 
> Everything now seems to be running! I will now have a look at the code specifics :)
> 
> Quick q about skypilot: Do we need to use jupyter notebooks? Could we instead just run the python files through the terminal we have SSH'd into?

Thanks so much Jess!

Re: the notebook issue: were you in jupyter, or were you using VSCode? In jupyter I just right-clicked the .py file and opened as a notebook.

I had the same issue with the kernel not existing and I just selected the Python 3 kernel 😬 What you did is more robust. I can try adding some lines to the yaml so that the kernel gets created at the setup stage.

And yeah I think we can also run scripts - no need to use notebooks! I was working in a notebook at this stage because I wanted to be able to tweak things and interactively look at the outputs.

**Comment by j-gillam** [comment]
> > Hey @RFOxbury ! Skypilot launched perfectly with your instructions!! It did ask me for the token.
> > I have had to do a couple of extra steps to get your script to run: It wouldn't let me open the py file as an ipynb. I instead had to convert from py to ipynb, e.g. in the terminal type: `jupytext --to notebook discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py` and also needed to add the kernel so it would show up as an option in jupyterlab: `python -m ipykernel install --user --name=discovery_child_development`
> > Everything now seems to be running! I will now have a look at the code specifics :)
> > Quick q about skypilot: Do we need to use jupyter notebooks? Could we instead just run the python files through the terminal we have SSH'd into?
> 
> Thanks so much Jess!
> 
> Re: the notebook issue: were you in jupyter, or were you using VSCode? In jupyter I just right-clicked the .py file and opened as a notebook.
> 
> I had the same issue with the kernel not existing and I just selected the Python 3 kernel 😬 What you did is more robust. I can try adding some lines to the yaml so that the kernel gets created at the setup stage.
> 
> And yeah I think we can also run scripts - no need to use notebooks! I was working in a notebook at this stage because I wanted to be able to tweak things and interactively look at the outputs.

I tried right clicking and opening with notebook and it came up with a big error when I tried to open it! :/ which I remembered having a year or so ago so just used the fix above! 

Ooo that would be really cool if it was in the yaml as well! If not, I think that extra step in the instructions is super easy to do! 

Just had a go at running a python script while I was waiting for the script to run, and it all seems to work well (and fast!). 

This is really great, thank you so much for setting it all up!!

File: infra/skypilot/README.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+Loosely based on [these docs](https://skypilot.readthedocs.io/en/latest/examples/gpu-jupyter.html#:~:text=SkyPilot%20makes%20the%20process%20of,managing%20provisioning%20and%20port%20forwarding.&text=View%20the%20supported%20GPUs%20with%20the%20sky%20show%2Dgpus%20command.&text=Enter%20the%20password%20or%20token,can%20create%20a%20new%20notebook.).
```
Comment (j-gillam) [inline infra/skypilot/README.md:1]: In the README instructions, using `exit` didn't seem to do anything. Any chance you could write separate instructions for each of the terminals? (i.e the one you type sky launch and the one you SSH into). 

Q: What is the different between sky down and sky stop? Options showed up when I pressed control+c to get out of the instance after `exit` didn't work.

File: infra/skypilot/task.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  git clone https://github.com/nestauk/discovery_child_development
+  cd discovery_child_development
+  git checkout 34-fine-tuning-and-skypilot #remove this line prior to merging to dev
+  direnv allow .
+  make install
+
+run: |
+  cd ~/sky_workdir/discovery_child_development
+  conda deactivate
+  conda activate discovery_child_development
```
Comment (j-gillam) [inline infra/skypilot/task.yaml:31]: Think you can just put the kernel code line in here, and it should work?

File: discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py [hunk 3]
```diff
@@ -319,5 +346,62 @@ def load_trainer(
...
+
+# %%
+# Log confusion matrix
+wb_confusion_matrix = wandb.Table(
+    data=confusion_matrix, columns=confusion_matrix.columns
+)
+run.log({"confusion_matrix": wb_confusion_matrix})
+
+# %%
+wandb.finish()
```
Comment (j-gillam) [inline discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py:381]: Was having a look at your previous wandb run and our accuracy is a little different? Is that to be expected with the hugging face models or should it be the same since you specified the seed? Unless you were running something different?

File: discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py [hunk 4]
```diff
@@ -18,8 +18,14 @@
 )
```
Comment (j-gillam) [inline discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py:18]: I was having a read about the tradeoffs between epochs and batch-size/ worrying about overfitting. I would also think there should be a few more epochs! Maybe we can just decide on a number? Alternatively, we could use some form of elbow method approach and run it for a lot of different number of epochs and see where it plateaus?

Comment (RFOxbury) [reply discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py:18]: I reckon in the short-term we could change it to the same hyperparameters Karlis used last time (in [this code](https://github.com/nestauk/innovation_sweet_spots/blob/dev/innovation_sweet_spots/analysis/notebooks/review_labelling/utils.py)) and then as a next step do a weights and biases sweep to find the best set of hyperparameters! 😱

I've been watching [this video](https://www.youtube.com/watch?v=6wRBGNLgQFU&list=PL--TX74VfiK1__SDFzriBXw5_a6rOOYqo&index=1)

Comment (RFOxbury) [reply discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py:18]: Elbow method is also a good thought and there may be a way to visualise that with weights and biases?

I think `EarlyStoppingCallback` kind of does the same thing as the elbow method, in that it stops the training if the fit has not improved for X epochs in a row

If you've read anything good, can you send me a link?

File: discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py (hunk 3 above)
Comment (RFOxbury) [reply discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py:381]: 🧐 no clue. Will have a google

File: infra/skypilot/task.yaml (hunk 2 above)
Comment (RFOxbury) [reply infra/skypilot/task.yaml:31]: Great, will do!

File: discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py [hunk 5]
```diff
@@ -44,9 +50,10 @@
 ## project code
 from discovery_child_development import PROJECT_DIR, logging, S3_BUCKET, config
 from discovery_child_development.getters import openalex as oa
-from discovery_child_development.pipeline.models import baseline_model as bm
+from discovery_child_development.pipeline.models.taxonomy_classifier import (
```
Comment (j-gillam) [inline discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py]: Think you might be able to delete this? I couldn't see where bm was used?

File: infra/skypilot/README.md (hunk 1 above)
Comment (RFOxbury) [reply infra/skypilot/README.md:1]: Oh interesting ok 🤔

Hm I think it's like the difference between putting your computer to sleep, and shutting your computer down. From [these docs](https://skypilot.readthedocs.io/en/latest/reference/cli.html?highlight=sky%20down#sky-stop) it seems like you can't use `sky stop` if you have a spot instance (which is what we're using because it's cheaper). `stop` preserves the data that was on the cluster, while `down` completely gets rid of the cluster and any data you had on it.

Comment (RFOxbury) [reply infra/skypilot/README.md:1]: And yes I'll write the instructions for the separate terminals 👍🏻

File: discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py (hunk 5 above)
Comment (RFOxbury) [reply discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py]: Good spot, yes that's not needed

File: discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py (hunk 4 above)
Comment (j-gillam) [reply discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py:18]: Ooo good shout!! I'm sure there must be a way to visualise with wandb! 

EarlyStoppingCallback looks super interesting!

File: discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py [hunk 6]
```diff
@@ -214,6 +217,14 @@ def load_trainer(
...
+run = wandb.init(
+    project="ISS supervised ML",
+    job_type="Taxonomy classifier",
+    save_code=True,
+    tags=["finetuning"],
+)
+
 # %%
 # Load the data. Just the training set by default
 openalex_data, training_file_name = oa.get_labelled_data()
```
Comment (j-gillam) [inline discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py:229]: How do the two different types of embeddings work? You load in 'embeddings' and join that into the openalex data. But then also use df_to_hf_ds, which also tokenizes the data? Is the second one used in the model?

**Review by j-gillam — APPROVED** [review]
Hey! Think this all looks really great, makes sense to me! And skypilot is working which is awesome!! Thanks so much for doing it, I'm super excited to use it!! 

Had a couple of questions (it wouldn't let me comment on some parts of the file) for 06_fine_tune_embeddings:
Does the num_labels parameter need to exist? Is there a use case for when the num_labels!=length of the trainingsets labels. I think the method has a default.. which might be that anyway? (Might be wrong!)

You may want to save the model to S3 as a pickle file rather than locally? As then it will just save to the skypilot instance?

File: discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py (hunk 6 above)
Comment (RFOxbury) [reply discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py:229]: Yes good spot! Loading in the embeddings was not needed. I think the tokenizing step creates the embeddings that are needed as input to the Distilbert training.

**Comment by beingkk** [comment]
Thanks @RFOxbury I will aim to review in the next couple of days!

File: discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py (hunk 6 above)
Comment (RFOxbury) [reply discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py:229]: Good question re: num_labels. I did not know the answer. [This tutorial](https://medium.com/analytics-vidhya/finetune-distilbert-for-multi-label-text-classsification-task-994eb448f94c) also supplies a parameter called `num_classes` so I think it is mandatory. Have you seen anything that makes you think there's a default?

I asked ChatGPT and it had a whole lot to say on the topic!
"""
In the HuggingFace Transformers library, the num_labels parameter in the AutoModelForSequenceClassification function is used to specify the number of distinct labels that the model should predict for a given task. This is particularly relevant for classification tasks, where the model needs to predict which category or class a given input belongs to.

When you're dealing with a classification problem, the model outputs a prediction for each label, and the num_labels parameter tells the model how many different output labels it needs to prepare for. This is crucial for the model to understand the shape and structure of the output layer, where the final classification decisions are made.

For instance, if you are training a sentiment analysis model where the input texts can be classified as either positive, negative, or neutral, you would set num_labels=3 since there are three possible categories.

In the context of the code snippet you provided, where problem_type is set to "multi_label_classification", the num_labels parameter becomes even more significant. It indicates that each instance in the dataset can belong to multiple categories at once, and the model needs to predict a set of labels, not just one. Each label in the output will be treated as a separate binary classification, determining whether or not the instance belongs to that label.
"""

Comment (j-gillam) [reply discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py:229]: I love that you asked chatgpt ahaha. I was having a look at the [source code](https://github.com/huggingface/transformers/blob/v4.35.2/src/transformers/configuration_utils.py#L49). You can just search for num_labels and it seems that if you don't specify it will go to 2? Which is definitely not what you wanted (but is what i want), so I think you have done it correctly! I think I may have just been misunderstanding what  `len(train_ds[0]["labels"])` was returning. Is that the maximum number of categories possible?

**Comment by beingkk** [comment]
Thanks so much for this @RFOxbury, great to see this. 
I managed to successfully run it (from Nesta office) and didn't run into any errors.

I can see that there's been a lot of discussion already, and don't have much to add at the moment. Perhaps one point is that the fine tuning notebook has lots of reusable functions/utils - but happy for this to be factored out as part of another PR.

Happy for you to merge!

**Review by beingkk — APPROVED** [review]
